Idl parser

seriyps opened the pull request:
Implementation of [idl](https://avro.apache.org/docs/++version++/idl-language/) parser. Fixes #68.


## Usage Examples

```
// file: my_service.avdl

@namespace("com.example")
protocol MyService {
  enum Status { OK, ERROR }

  record Response {
    string id;
    Status status;
    union { null, string } message = null;
  }
}
```

### Load a schema file and encode/decode

```erlang
Store = avro_schema_store:new([], ["my_service.avdl"]),
LookupFun = avro_schema_store:to_lookup_fun(Store),
Encoder = avro:make_encoder(LookupFun, []),
Decoder = avro:make_decoder(LookupFun, []),

Record = [{"id", <<"req-1">>}, {"status", <<"OK">>}],
Bin = iolist_to_binary(Encoder("com.example.Response", Record)),
Record = Decoder("com.example.Response", Bin).
```

### Convert .avdl to an AVPR map (JSON protocol representation)

```erlang
{ok, Bin} = file:read_file("my_service.avdl"),
Avpr = avro_idl:str_to_avpr(
         binary_to_list(Bin),
         filename:dirname("my_service.avdl")),
%% Avpr is a map: #{<<"protocol">> => ..., <<"types">> => [...], <<"messages">> => ...}
```

### In-memory schemas (useful for tests or embedded schemas)

```erlang
Files = #{
    {"schemas", "common.avdl"} =>
        <<"protocol Common { record Address { string city; } }">>,
    {"schemas", "order.avdl"} =>
        <<"protocol Orders {  import idl \"common.avdl\"; record Order { string id; int amount; } }">>
},
ReadFun = fun(Cwd, Path) ->
    case maps:find({Cwd, Path}, Files) of
        {ok, Bin} -> {ok, Bin};
        error     -> {error, enoent}
    end
end,
{ok, Src} = maps:find({"schemas", "order.avdl"}, Files),
Schema = avro_idl:decode_schema(
           binary_to_list(Src), "schemas",
           [{read_fun, ReadFun}]).
```

seriyps committed:
.avdl lexer+parser draft

seriyps committed:
More progress on parser

Seems the only things left are some of annotations and docstrings

seriyps committed:
More strict lexical rule for namespaced ID

seriyps commented on src/avro_idl_lexer.xrl:
```diff
@@ -0,0 +1,85 @@
+
+
+unescape(Token, Char) ->
+    string:trim(Token, both, [Char]).
```
can be replaced with `string:strip` for OTP 19 compatibility

seriyps committed:
Use `string:strip` to be compatible with OTP19

seriyps committed:
Don't ignore function return type

seriyps committed:
Full annotations support; parser tests added

seriyps committed:
Add support for docstrings; rename `annotations` field to `meta`

seriyps committed:
IDL to avpr/avsc converter; support multiple `throw` types

seriyps committed:
Better test coverage

seriyps committed:
Optimize doc strip preprocessor by pre-compiling regexp

seriyps committed:
Allow any type inside array and map; fix single-letter namespaced IDs

seriyps committed:
Use binaries for avpr JSON-map representation

seriyps committed:
Load .avdl to erlavro internal format

Now it's possible to use basic .avdl to encode/decode avro

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/klarna/erlavro?pullRequest=90) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/klarna/erlavro?pullRequest=90) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/klarna/erlavro?pullRequest=90) it.</sub>

seriyps committed:
Fix lexer and parser correctness for IDL

Lexer fixes:
- Support empty string literals ("") — previously caused illegal token error
- Support backslash escape sequences in strings (\", \\, \n, \t, \r)

Parser fixes:
- Allow annotations and multiple doc comments before function/message
  definitions, consistent with how enum/record/error/fixed declarations
  are handled (meta function rule instead of doc_v function)
- Allow empty record and error type bodies (record Foo {})
- The meta function fix also resolved the undeclared 3rd shift/reduce
  conflict, so Expect 2 is now correct

seriyps committed:
feat: implement import resolution in avro_idl (idl/protocol/schema)

- process_imports/2 now loads all three import types instead of silently
  dropping them; imported types are returned as {avsc, Map} tagged tuples
  so they pass through protocol_to_avpr/2 without re-conversion
- load_idl_import/2 uses the imported file's dirname as cwd so nested
  relative imports resolve correctly
- Add tests for each import type and for nested directory imports

seriyps committed:
Add .avdl support to avro_schema_store:import_file/2

Detect ".avdl" extension and delegate to avro_idl:decode_schema/2,
then flatten the resulting union type and add each named type to the
store individually via do_add_type/2.

seriyps committed:
Add read_fun option for in-memory schema loading

Add a {read_fun, fun((Cwd, Path) -> {ok, Bin} | {error, term()})}
option to str_to_avpr/3 and decode_schema/3. The default read_fun
uses file:read_file/1. The fun is threaded through recursive IDL
imports so all file I/O can be intercepted at every level.

Also add decode_schema/3 as a convenience wrapper accepting the
same options.

seriyps committed:
docs: add Avro IDL section to README with usage examples

seriyps committed:
test: add encode/decode tests for .avdl schemas with imports

- Add encode_decode_test/0 in avro_idl_tests covering:
  - types from imported .avdl (idl import)
  - types from imported .avpr (protocol import)
  - types from imported .avsc (schema import)
  - cross-namespace type references (org.erlang.ftp.MyAnnotated)
- Fix protocol_with_typedefs.avdl: MyAnnotated referenced MyError
  (an error type filtered out by decode_schema), replaced with a
  valid reference to org.erlang.www.MyEnum2

seriyps commented:
@zmstone this is now ready. It took me some time to finalize it 😄

mikpe commented on src/idl.hrl:
```diff
@@ -0,0 +1,45 @@
+         meta = [],
+         fields = []}).
+
+-record(record,
```
Naming the record `record` causes problems with OTP-29. We _can_ silence them for a while using a compiler option, but it would be better to change the name before merging this.
I know it sounds lame, but `record_` should work.

seriyps replied:
we can probably add `idl_` prefix to all the records in this file

seriyps replied:
fixed, I renamed the records adding `idl_` prefix

mikpe replied:
Sorry, I was being incorrect. OTP-29 lets you name a record `record` no problem, but it will not let you name a _type_ `record`.
But your code is much nicer with that `idl_` prefix :-)

mikpe reviewed (approved):
With the caveat wrt the `record` name.

seriyps committed:
Rename idl records with idl_ prefix

The `record` would be a reserved word and would warn in OTP-29; the
rest of the records are updated for consistency

zmstone commented on rebar.config:
```diff
@@ -1,9 +1,9 @@
 %% -*- mode:erlang -*-
 {erl_opts,             [ debug_info
-                       , warnings_as_errors
+%                       , warnings_as_errors
```
bring this back ?

seriyps replied:
done

seriyps committed:
revert temporary changes in rebar.config
